logic/scorer/scorer.go

```diff
@@ -60,260 +34,247 @@ func points(lead model.Card, hand []model.Card, isCrib bool) int {
-	} else if values[2] == max {
-		// A|B|D D D
-		// A A|D D D
-		if values[1] == min {
-			return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-		}
-		return triplet, 6 /* triplet */
+func (fs *flushScorer) accumulate() (scoreType, uint8) {
+	if !fs.isHandFlush {
+		return 0, 0
+	}
+	if fs.firstCard.Suit == fs.leadCard.Suit {
+		return flush5, 5
+	}
+	if fs.isCrib {
+		return 0, 0
 	}
+	return flush4, 4
+}
 
-	switch diffOf5Cards {
-	case 3:
-		// If we have a diff of 3, then we either have a two pair or a double run of 4
-		// check double run of four: A A|B|C|D
-		// check two-pair          : A A|B B|D
-		numPairs := 0
-		for i := 1; i < len(values); i++ {
-			if values[i] == values[i-1] {
-				numPairs++
-			}
-		}
+type iterateHandResult struct {
+	valuesToCounts [valuesToCountsLength]uint8
+	values         [numCardsToScore]int
+	ptValues       [numCardsToScore]int
+	scoreType      scoreType
+	totalPoints    uint8
+}
 
-		if numPairs == 2 {
-			// we found two pairs, which mean it cannot be a double run of 4
-			return twopair, 4 /* two pair */
+// iterateHand does all the scoring possible with a single iteration through the hand
+// we can't do fifteens or runs here directly, but the information returned will aid in doing those efficiently
+func iterateHand(lead model.Card, hand []model.Card, isCrib bool) iterateHandResult {
+	res := iterateHandResult{}
+	hasNobs := false
+	ps := &pairScorer{}
+	fs := newFlushScorer(lead, hand, isCrib)
+	for i, c := range [5]model.Card{hand[0], hand[1], hand[2], hand[3], lead} {
+		ps.atEach(c)
+		fs.atEach(c)
+		if c != lead && c.Value == model.JackValue && c.Suit == lead.Suit {
+			hasNobs = true
 		}
+		res.values[i] = c.Value
+		res.ptValues[i] = c.PegValue()
+	}
 
-		return doubleRunOfFour, 10 /* double run of four */
+	if hasNobs {
+		res.totalPoints++
+		res.scoreType |= nobs
+	}
+	fSt, fPts := fs.accumulate()
+	res.totalPoints += fPts
+	res.scoreType |= fSt
 
-	case 4:
-		// So here's the options:
-		// Run of 5
-		// 1,2,3,4,5
-		// double run of three
-		// 1,2,x,3,5
-		// two pair
-		// 1,1,3,3,5
-		// 1,3,3,5,5
-		// pair + run of three
-		// 1,2,3,5,5
-		// 1,1,3,4,5
-		// pair alone (regardless of fifteens)
-		// 6,6,7,9,10
-		// 6,7,9,10,10
-		var numPairs, numIncs int
-		for i := 0; i < len(values)-1; i++ {
-			if iv, nv := values[i], values[i+1]; iv == nv {
-				numPairs++
-			} else if iv+1 == nv {
-				numIncs++
-			}
-		}
-		if numPairs == 2 {
-			return twopair, 4 /* two pair */
-		} else if numIncs == 4 {
-			return run5, 5 /* run of 5 */
-		} else if values[2] == max-2 && (values[3] == max || values[1] == min) {
-			return run3 | onepair, 5 /* run of three and a pair*/
-		} else if (values[2] == min+1 && values[3] != values[2]+1) ||
-			(values[2] == max-1 && values[1] != values[2]-1) {
-			return onepair, 2 /* a pair*/
-		}
+	pSt, pPts := ps.accumulate()
+	res.totalPoints += pPts
+	res.scoreType |= pSt
+	res.valuesToCounts = ps.valuesToCounts
```

Comment:
:cactus: now the `pairScorer` has two responsibilities, breaking S of SOLID (Singe responsibility principle).

the first responsibility is to score pairs.
the second is to provide a mapping of card values to number of that value.

Reply:
This is true - or, the map could be seen as a byproduct of scoring pairs that happens to be useful for scoring runs
The alternative is to generate the map in `iterateHand` and pass it into `atEach`; this would result in copying it 5 times. Which would you rather do?